tmt/steps/execute/internal.py

```diff
@@ -192,45 +192,45 @@ def execute(self, test: Test, guest: Guest,
 
         test_wrapper_filepath = workdir / TEST_WRAPPER_FILENAME
 
-        # Prepare the test command (use default options for shell tests)
-        if test.framework == "shell":
-            test_command = ShellScript(f"{tmt.utils.SHELL_OPTIONS}; {test.test}")
-        else:
-            test_command = test.test
-        self.debug('Test script', str(test_command), level=3)
-
-        # Prepare the wrapper, push to guest
-        self.write(test_wrapper_filepath, str(test_command), 'w')
-        test_wrapper_filepath.chmod(0o755)
-        guest.push(
-            source=test_wrapper_filepath,
-            destination=test_wrapper_filepath,
-            options=["-s", "-p", "--chmod=755"])
-
-        # Prepare the actual remote command
-        remote_command = ShellScript(f'./{TEST_WRAPPER_FILENAME}')
-        if self.get('interactive'):
-            remote_command = ShellScript(
-                TEST_WRAPPER_INTERACTIVE.format(
-                    remote_command=remote_command))
-        else:
-            remote_command = ShellScript(
-                TEST_WRAPPER_NONINTERACTIVE.format(
-                    remote_command=remote_command))
-
-        # Execute the test, save the output and return code
-        start = time.time()
-        try:
-            stdout, _ = guest.execute(
-                remote_command,
+        def run_cmd(cmd: ShellScript) -> tmt.utils.CommandOutput:
+            # Prepare the wrapper, push to guest
+            self.write(test_wrapper_filepath, str(cmd), 'w')
+            test_wrapper_filepath.chmod(0o755)
+            guest.push(
+                source=test_wrapper_filepath,
+                destination=test_wrapper_filepath,
+                options=["-s", "-p", "--chmod=755"])
+            if test.framework == "shell":
+                cmd = ShellScript(f"{tmt.utils.SHELL_OPTIONS}; {cmd}")
+            self.debug('Test script', str(cmd), level=3)
+            if self.get('interactive'):
+                cmd = ShellScript(
+                    TEST_WRAPPER_INTERACTIVE.format(remote_command=cmd))
+            else:
+                cmd = ShellScript(
+                    TEST_WRAPPER_NONINTERACTIVE.format(remote_command=cmd))
+            return guest.execute(
+                cmd,
                 cwd=workdir,
                 env=environment,
                 join=True,
                 interactive=self.get('interactive'),
                 log=self._test_output_logger,
                 timeout=tmt.utils.duration_to_seconds(test.duration),
```

Comment:
The duration should apply to the test as a whole - with `duration: 5m` & a setup test that takes an hour, we could easily run out of time after setup already. I'd say we need to establish a deadline and pass "remaining time" to `timeout`.

Reply:
I've added a static local variable `run_cmd.spent_time`, which tracks how much time is already spent on the setup/test/cleanup execution. What do you think about this?

Reply:
Looks good, although I'm afraid this won't survive in the long run :/ In https://github.com/teemtee/tmt/pull/2210 I'm adding something very similar, I'm refactoring the code you're moving around, and just like your setup/cleanup, before & after test checks will need to be constrained by a deadline, and we will have to store their timestamps in results, and so on. Some kind of reusable context manager or something, to manage the decreasing time left automagically for us...

This will be a PITA to join :/